### PR TITLE
Tweaks 7

### PR DESCRIPTION
**Pull request by mohawk2**
Further simplifications towards native complex.

### CODE REVIEW

**Comment by mohawk2** [comment]
More questions! In MHProd, the current test I have gives the metric as a real `ones(1, 1, 11)`, which I *think* is being treated as a complex with both real and imaginary parts as 1 (thanks to threading). The closest I have to making it work with native complex has changed it from returning a 1-dimensional ndarray to a zero-dimensional one, which I think is a sign I'm getting something wrong. The adding of a dummy dimension makes me unsure what's happening.

Basically, what should the dimensional relationship between the two fields and the metric be? Are they all three supposed to be complex?

**Comment by mohawk2** [comment]
SProd appears to be reversing all its dimensions including switching real parts with imaginary, is that intentional? (I'm not yet fully understanding spinors, sorry)

**Comment by mohawk2** [comment]
GtoR and RtoG are only actually used in ::LE::NR2::SH.

Both take a `$skip` parameter which blindly moves dimensions to the end, to be threaded over, possibly including the 0th "RorI" one. I want to change them so they only skip non-complex dimensions, and encapsulate this inside the two functions, and change the calls to them. This will hide more detail about the complex number stuff.

This will be fine for `externalVecL`, `selfConsistentVecL`, `external_G` which call them with 1 skip, which is what I'd expect, and would then change to 0.

However, `P2LMCalt` calls them both with 0 skip (in calculating `$beta_G`), i.e. threading over the "RorI" dimension. The doc for RtoG etc say "skip the next dimensions [after the first, complex one]" so I will change that skip to 1 for both of them, and adjust the test results, based on my understanding that this method hasn't really been tested/used yet.

**Comment by wlmb** [comment]
> More questions! In MHProd, the current test I have gives the metric as a real `ones(1, 1, 11)`, which I _think_ is being treated as a complex with both real and imaginary parts as 1 (thanks to threading).

The problem we solved by introducing the metric had a real (not complex) metric g that is a cartesian tensor with components g_xx g_xy g_yx g_yy... Thus, for D dimensional space, the first two dimensions are D. Furthermore, there is one of this tensors for each wavevector G, which has components G_x G_y... Thus, the next D dimensions are the possible values for each component of G: N_x N_y... These N's are equal to the number of pixels (or voxels) used to define the characteriztic function. 

In the test, D=1 and N_x=11, so the metric is a 1,1,11 real pdl.

> The closest I have to making it work with native complex has changed it from returning a 1-dimensional ndarray to a zero-dimensional one, which I think is a sign I'm getting something wrong. The adding of a dummy dimension makes me unsure what's happening.
> 
> Basically, what should the dimensional relationship between the two fields and the metric be? Are they all three supposed to be complex?

With a skip argument things would be slightly more complicated than  what I said above, but it is currently disallowed and I haven't used that parameter yet. 

Let me consider a 3D example with 11x11x11 voxels:
The first and second states would be complex with dimensions ri,D,nx,ny,nz=2,3,11,11,11
$ndim would then be 5
The metric would have dimensions D,D,nx,ny,nz=3,3,nx,ny,nz
Then $mprod would have dimensions  ri,D,nx,ny,nz 
as the states
$prod will also have dimensions ri,D,nx,ny,nz
The reorder in line 155 would read reorder(1,2,3,4,0) resulting in dimensions D,nx,ny,nz,ri
This intermediate result is disagreable as I moved ri to the end, so it is not a well formed PDL::Complex, a reason for migrating to native complexes. 
The clump would be clump(-2) resulting in dimensions D*nx*ny*nz,ri=3993,2
The mv(-1,0) restores the correct format ri,D*nx*ny*nz
The sumover then sums over dimension 1 (as sumover recognizes PDL::Complex's) and yields a single complex number. 

Hope this helps.

**Comment by wlmb** [comment]
> SProd appears to be reversing all its dimensions including switching real parts with imaginary, is that intentional? (I'm not yet fully understanding spinors, sorry)

I should explain about the spinors somewhere. And I should also write about Photonic somewhere. I haven't given myself the time.  I developed the spinor code about two years ago and hadn't used it much, though I like it as it allows applying my theory to arbitrary metamaterials with any kind of dissipative response. The main idea is that when you have many materials and they are dissipative, the response of the system is not Hermitian so you cannot use the usual theorems about Hermitian matrices, but when you consider simultaneously two waves with opposite directions, the response becomes symmetric, and the theorems for symmetric matrices are similar enough to those of Hermitian matrices for our purposes. If you are curious, take a look at https://doi.org/10.1002/pssb.201900560

Anyway, going to your question: I do reverse a lot of dimensions, but not the real-imaginary dimension. Look at line 199 which says 'keep the complex dimension' by setting the first slice argument to ':'.

**Comment by mohawk2** [comment]
Thank you for taking the time to answer my ill-informed questions! The above all makes a great deal of sense. Regarding MHProd, the extra dimension of PDL::Complex confused me yet again - the zero-dim return value exactly matches the single-dim=2 of the PDL::Complex before my modifications, so all is well there.

I have actually had a read of "Macroscopic optical response and photonic bands", and actually did understand small portions of it. I shall look at your paper linked above in due course, thank you.

Locally I actually have fully modified all the ::Utils functions, and `t/utils.t`, to be native-complex only, so the end is in sight!

**Comment by wlmb** [comment]
> Thank you for taking the time to answer my ill-informed questions! 

They are good questions and trying to answer refreshes my memory. 

>The above all makes a great deal of sense. Regarding MHProd, the extra dimension of PDL::Complex confused me yet again - the zero-dim return value exactly matches the single-dim=2 of the PDL::Complex before my modifications, so all is well there.
> 
> I have actually had a read of "Macroscopic optical response and photonic bands", and actually did understand small portions of it. I shall look at your paper linked above in due course, thank you.
> 
> Locally I actually have fully modified all the ::Utils functions, and `t/utils.t`, to be native-complex only, so the end is in sight!

Great! Thanks! I'll look later at your latest suggested changes. Best regards, L

File: t/oneh-np.t [hunk 1]
```diff
@@ -54,7 +54,7 @@ $o->iterate;
 ok(Cagree(pdl($o->current_a), $ea*$f+$eb*(1-$f)), "1D a_1");
 ok(Cagree(pdl($o->next_b2), 0), "1D b_2^2");
 my $x = zeroes(1, 2, 11)->r2C;
-$x->slice(':,:,0') .= r2C(1);
+$x->slice(':,:,0') .= 1;
```
[the remark this answers is not in the document]
Comment (wlmb) [reply t/oneh-np.t:57]: I've updated now, and it works as intended
